corehq/project_limits/gauge.py

```diff
@@ -0,0 +1,188 @@
+
+    def get_last_reported_time(self):
+        "returns the last time when the lag was reported or will return ``datetime.min`` if not reported yet"
+        return cache.get(self._last_reported_timestamp_key, datetime.min)
+
+    def max(self):
+        """
+        return max lag for a given topic
+        """
+        values = self.get_values()
+        return max(values) if values else None
+
+    def min(self):
+        """
+        return min lag for a given topic
+        """
+        values = self.get_values()
+        return min(values) if values else None
+
+    def avg(self):
+        """
+        return average lag for a given topic
+        """
+        values = self.get_values()
+        return sum(values) / len(values) if values else None
+
+
+case_pillow_lag_gauge = Gauge(feature_key=CASE_SQL, scope_fn=get_all_kafka_partitons_for_topic)
+
+
+class GaugeLimiter:
+
+    def __init__(self, gauge, get_gauge_definition):
+        """
+        :param gauge: ``Gauge`` An instance of Gauge class
+        :param get_gauge_defintions: ``function``
+            A function to get config that GaugeLimiter will use to
+            figure out throttling conditions.
+        """
+        self.gauge = gauge
+        self.gauge_definition_fn = get_gauge_definition
+
+    def wait(self):
+        if not self.allow_usage():
+            time.sleep(self.gauge_definition.throttle_for_seconds)
+
+    def allow_usage(self):
+        """
+        Checks for the throttle conditions provided in the Defintion class and evaluates them.
+        returns `True` it means that the usage should not be throttled.
+                `False` it means that the usage should be throttled.
+        """
+        if not self.gauge_definition:
+            # no throttling if config to throttle is not set.
+            return True
+        computed_value = None
+        if self.gauge_definition.actionable_metric == AVG:
+            computed_value = self.gauge.avg()
+        else:
+            computed_value = self.gauge.max()
+        return computed_value < self.gauge_definition.acceptable_value
+
+    @property
+    def gauge_definition(self):
+        """
+        Should return an instace of subclass of ThrottleDefinition
+        """
+        return self.gauge_definition_fn(self.gauge.feature_key)
+
+
+class PillowLagGaugeLimiter(GaugeLimiter):
+
+    def _has_pillow_reported_recently(self):
+        time_window = 15  # time window in minutes in which throttling should be applied
+        time_before_window = datetime.now(tz=timezone.utc) - timedelta(minutes=time_window)
+        return self.gauge.get_last_reported_time() > time_before_window
+
+    def allow_usage(self):
+        if not self.gauge_definition:
+            # No throttling if config to throttle is not set
+            return True
+        return not (self._is_ideal_throttle_condition() and self._has_pillow_reported_recently())
+
+    def _is_ideal_throttle_condition(self):
```

Comment:
Should this method make use of the gauge definition's aggregator type? It looks like it is defaulting to checking both average and max? Which maybe renders the aggregator unused?

Reply:
As per my conversation with Danny we would like to have a different throttle condition for the pillows where we are saying that it should breach both average and max values, so we have a overridden the `allow_usage` but base class uses `aggregator`

Reply:
Ahh I see. I couldn't find aggregator used elsewhere but I see that it was still referenced as `actionable_metric` in the base class. Left a comment on that.

Reply:
Yeah. I saw it when I was going through the code. Have fixed it in the last commit 😅